### PR TITLE
Implement Game, GameCreator, BoardCreator classes

### LINKED ISSUE #1: Initialize a Game
Creating a basic 8x8 chess board to start building more complicated systems.

### PR DESCRIPTION
**Pull request by cherche**
Also partially implements Game and Board

This is based on an older commit. I will rebase and correct discrepancies before merging. Only resolves #1—not #16.

### CODE REVIEW

**Comment by cherche** [comment]
Requesting documentation review by @Gursach44 and test coverage review from @RMagill46 @cjwssb687, before proceeding.

**Comment by cherche** [comment]
By the way, GitHub autograding is, for whatever reason, failing the tests. Cannot reproduce locally. Running `gradle test` results in passes.

Edit: I was wrong! Everything is working fine.

File: src/main/java/GameCreator.java [hunk 1]
```diff
@@ -11,7 +11,11 @@ public GameCreator() {}
     public Game create() {
         Game game = new Game();
 
-        // TODO: Rest of initialization
+        Player p1 = new Player("Player 1", 'W');
```
Comment (Gursach44) [inline src/main/java/GameCreator.java]: Should initialize players with the names provided when GameCreator is instantiated in the main method (wherever that is going)
   - We also have to add the main class to our framework #17

**Review by David9Nagy — DISMISSED** [review]
Looks good. We will need to add documentation for classes eventually.

**Review by Gursach44 — CHANGES_REQUESTED** [review]
Reviewed documentation. Looks good - we don't need to document much in terms of initialization I feel, because it's pretty easy to read.

**Review by Gursach44 — CHANGES_REQUESTED** [review]
Current documentation looks good, but need to add class documentation as well.

**Review by adirao-projects — APPROVED** [review]
Everything seems to look fine. There may be some future issues with `getTypeFromChar()` as I ran into some issues earlier in the CLIOutput class. It doesn't seem like a significant enough issue to not approve this change. Be aware that this may cause conflicts in the future.

**Comment by cherche** [comment]
> Everything seems to look fine. There may be some future issues with `getTypeFromChar()` as I ran into some issues earlier in the CLIOutput class. It doesn't seem like a significant enough issue to not approve this change. Be aware that this may cause conflicts in the future.

@adirao-projects what issues do you anticipate from the current implementation of `getTypeFromChar()`? It relies on no other classes (besides `PieceFactory.java`, in which the piece type strings  `"pawn"`, `"knight"`, etc. were chosen). It is my understanding that the issue with `CLIOutput.java` was that it relied on `Piece.java`, expecting some interface which would return a `char` representation of the piece. All of the `char` representations in this commit are totally internal to `BoardCreator.java`; they only exist to improve readability.

In general, I think it best that `char` representations be kept outside of the `Piece` class. They seem to exist only for the sake of writing slightly cleaner code. In `CLIOutput.java` for instance, every use of a `char` shorthand representation could very well be replaced with checks for the piece type string and the piece's player's colour.

**Review by David9Nagy — CHANGES_REQUESTED** [review]
in Board class:
- Do `containsAbsCoords`, `containsMirrorCoords`, and `ContainsRelativeCoords` do the same thing? Don't they just check if the coordinates are in between 0 and 7 inclusive?
- I like this relative, mirror, and absolute board model. Making sure I understand it correctly, mutating one type of board mutates the other two due to aliasing right?
- Should we add a method that tells us what the current POV of relative board is or should we just use the selectPOV methods everytime we edit relative board to ensure we're on the correct POV?

in BoardCreator class:
- From `boardString` it looks like lowercase characters are white and uppercase are black. Can you include this in the documentation? Method `createPieceFromChar` is written as if player p1 is black and p2 is white (if upper case player = p1 and vice versa) but doc says p1 is white. I think the if statements need to be switched.
- method `placeAllPieces` looks like it works for any board string layout. If so shouldn't you be able to pass in the `boardString` variable so that you can create a board instance from any position? If this is the case, the doc should be updated as well.
- less important suggestion: isn't `getTypeFromChar` redundant? Can't you figure out what type a piece should be from its character? Probably won't be worth changing anyways though as I see PieceFactory is already implemented to read the type as a string

**Comment by cherche** [comment]
> in Board class:
> 
>     * Do `containsAbsCoords`, `containsMirrorCoords`, and `ContainsRelativeCoords` do the same thing? Don't they just check if the coordinates are in between 0 and 7 inclusive?
> 
>     * I like this relative, mirror, and absolute board model. Making sure I understand it correctly, mutating one type of board mutates the other two due to aliasing right?
> 
>     * Should we add a method that tells us what the current POV of relative board is or should we just use the selectPOV methods everytime we edit relative board to ensure we're on the correct POV?
> 
> 
> in BoardCreator class:
> 
>     * From `boardString` it looks like lowercase characters are white and uppercase are black. Can you include this in the documentation? Method `createPieceFromChar` is written as if player p1 is black and p2 is white (if upper case player = p1 and vice versa) but doc says p1 is white. I think the if statements need to be switched.
> 
>     * method `placeAllPieces` looks like it works for any board string layout. If so shouldn't you be able to pass in the `boardString` variable so that you can create a board instance from any position? If this is the case, the doc should be updated as well.
> 
>     * less important suggestion: isn't `getTypeFromChar` redundant? Can't you figure out what type a piece should be from its character? Probably won't be worth changing anyways though as I see PieceFactory is already implemented to read the type as a string

In Board class:

* Actually, yes, they happen to do the same thing because we are on a rectangular board. The relative coordinates interface is rather clunky.
* Yes, that's pretty much right. I am glad you appreciate it! However, all of this business with 2d arrays is just an implementation quirk, and could be readily reimplemented so as not to use `board`, `mirror`, and `relative` (and, in fact, could even not use 2d arrays at all!).
* The intention is to use `selectPov()` before every use of relative coordinate methods. Practically, `getPov()` would not be particularly useful; there is never a case in which one would use `getPov()` without doing some check and ultimately calling `selectPov()` if the pov is not what is desired. Although, most of these coordinate manipulation methods should be moved, for they clutter the `Board` class. Correct me if I'm wrong, but it seems that only `MovesGenerator` and its subclasses need to use these methods.

In BoardCreator:

* No, the uppercase letters are white and the lowercase letters are black, as one sees from my passed tests. The confusion may lie with how the board appears when printed as a string, compared to how one usually sees a chess board when playing online. The first character in the board string has coordinates (0,0), the second has coordinates (1,0), etc. Consequently, there is a white rook at (0,0), a white knight at (0,1), etc., exactly as we discussed previously. Even if this were not the case, the use of chars to represent pieces is totally internal to the class; this interface is not publicly exposed and could never be used by another class.
* This is not the case; BoardCreator exists only to create a brand new board, as written in the class javadoc. If one wishes to extend BoardCreator to create a board from a board string, they should do so in a separate pull request.
* I don't know what you mean by "Can't you figure out what type a piece should be from its character?" Pieces do not have characters. Each Piece instance has a type which is a string, e.g., `"pawn"`, `"knight"`, and so on. Both PieceFactory and Piece understand types to be strings. So no, `getTypeFromChar()` is not redundant; a priori, there is no way to conclude that characters `'p'` or `'K'` correspond to the types `"pawn"` and `"king"`.

I shall summarize the above discussion. The `Board` class is a tad large with all its coordinate manipulation methods, but is not within the scope of this pull request. The `BoardCreator` class behaves exactly as documented. I would like clarification on your last point on `getTypeFromChar()`.

**Review by Gursach44 — APPROVED** [review]
Changes look good to me!

**Comment by David9Nagy** [comment]
I agree with all the points. I got confused regarding the layout of boardString as you said. I'll set it as approved.

Regarding the second last bullet point, the actual method placeAllPieces looks like it could work with any board layout, so if we're still planning to implment the ability to undo moves we can use it for that. This might be too early to be thinking about this though as we should aim to have a working game first and then add features.

last bullet point: I meant that we could change it so that the types are represented by characters instead of strings so every 'king' would be replaced by char 'k'. It doesn't really do much though so let's leave it as is. Comments in getType method in Piece class should be changed accordingly though, as it's currently
`    public String getType() {
        //Returns P for pawn, B for bishop, N for knight, 
        //R for rook, Q for queen, K for king
        return this.type;
    }
`

**Review by David9Nagy — APPROVED** [review]

**Comment by cherche** [comment]
> 

Yes, that comment was written by @adirao-projects when there was still uncertainty as to the interface for the Piece class. I do think it may be very helpful to extract a method which creates a board from a board string, for the sake of usability, readability, and easier testing.

**Pull request merged**